cache/filecache/filecache.go

```diff
@@ -0,0 +1,141 @@
+	// if not present (to separate prefix from the rest of the
+	// cache key).
+	FilenamePrefix string
+	// Defaults to ".bin" if left empty. This is added to the
+	// cache key and prefix above to form a file name.
+	FilenameExtension string
+	// The mode to use when creating new files in the cache
+	// directory. Defaults to 0600 if left zero. If not left
+	// as default, the mode must have at least bits 0400 and
+	// 0200 (read and write permissions for owner) set.
+	FileMode fs.FileMode
+}
+
+// New creates a new cache that stores items in files. The given
+// path must be a directory, in which the files will be stored.
+func New(config Config) (prototransform.Cache, error) {
+	// validate config
+	if config.Path == "" {
+		return nil, errors.New("path cannot be empty")
+	}
+	path, err := filepath.Abs(config.Path)
+	if err != nil {
+		return nil, err
+	}
+	config.Path = path
+	if config.FilenamePrefix == "" {
+		config.FilenamePrefix = "cache"
+	} else {
+		config.FilenamePrefix = strings.TrimSuffix(config.FilenamePrefix, "_")
+	}
+	if config.FilenameExtension == "" {
+		config.FilenameExtension = ".bin"
+	} else if !strings.HasPrefix(config.FilenameExtension, ".") {
+		config.FilenameExtension = "." + config.FilenameExtension
+	}
+	if config.FileMode == 0 {
+		config.FileMode = 0600
+	} else if (config.FileMode & 0600) != 0600 {
+		return nil, fmt.Errorf("mode %#o must include bits 0600", config.FileMode)
+	}
+
+	//  make sure we can write files to cache directory
+	info, err := os.Stat(path)
+	if err != nil {
+		return nil, err
+	}
+	if !info.IsDir() {
+		return nil, fmt.Errorf("%s is not a directory", path)
+	}
+	testFile := filepath.Join(path, ".test")
+	f, err := os.OpenFile(testFile, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0600)
+	if err != nil {
+		if os.IsPermission(err) {
+			return nil, fmt.Errorf("insufficient permission to create file in %s", path)
+		}
+		return nil, fmt.Errorf("failed to create file in %s: %w", path, err)
+	}
+	closeErr := f.Close()
+	rmErr := os.Remove(testFile)
+	if closeErr != nil {
+		return nil, closeErr
+	} else if rmErr != nil {
+		return nil, rmErr
+	}
+
+	return (*cache)(&config), nil
+}
+
+type cache Config
+
+func (c *cache) Load(_ context.Context, key string) ([]byte, error) {
+	fileName := filepath.Join(c.Path, c.fileNameForKey(key))
+	return os.ReadFile(fileName)
+}
+
+func (c *cache) Save(_ context.Context, key string, data []byte) error {
+	fileName := filepath.Join(c.Path, c.fileNameForKey(key))
+	return os.WriteFile(fileName, data, c.FileMode)
+}
+
+func (c *cache) fileNameForKey(key string) string {
+	if key != "" {
+		key = "_" + sanitize(key)
+	}
+	return c.FilenamePrefix + key + c.FilenameExtension
+}
+
+func sanitize(s string) string {
+	var sb strings.Builder
+	for _, r := range s {
+		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') ||
+			(r >= '0' && r <= '9') || r == '.' || r == '-' || r == '_' {
+			sb.WriteRune(r)
+			continue
+		}
+		sb.WriteRune('_')
```

Comment:
Won't this create conflicts why not base64/32 encode it?

Reply:
You are correct that this naive approach could create conflicts. (This PR also handles that so that it's not possible for to read the wrong schema from the cache, since the cache entry itself also includes the schema identity; but there still could be cache-thrashing on a collision, where the conflicting workloads constantly overwrite each others' cache.)

One issue with an encoding like base64 or crockford it is that it makes hard to tell what file is what. It seemed like a nice property for debugging that the file name would have the key in it and be readable.

How about something more like URL-encoding: instead of replacing everything with underscore, we replace everything (including percent) with a percent followed by 2-digit hex value. That would allow it to be conflict-free.

Reply:
Done.